torch/testing/_internal/autograd_function_db.py

```diff
@@ -0,0 +1,250 @@
+        ctx.mark_non_differentiable(ind, ind_inv)
+        ctx.save_for_backward(ind, ind_inv)
+        ctx.dim = dim
+
+    @staticmethod
+    def backward(ctx, grad_output, _0, _1):
+        ind, ind_inv = ctx.saved_tensors
+        return NumpyTake.apply(grad_output, ind_inv, ind, ctx.dim), None
+
+
+def sample_inputs_numpy_sort(opinfo, device, dtype, requires_grad, **kwargs):
+    make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+    yield SampleInput(make_arg(3, 5), args=(1,))
+
+
+class NumpyTake(torch.autograd.Function):
+    @staticmethod
+    def forward(x, ind, ind_inv, dim):
+        device = x.device
+        x = to_numpy(x)
+        ind = to_numpy(ind)
+        return torch.tensor(np.take_along_axis(x, ind, dim), device=device)
+
+    @staticmethod
+    def setup_context(ctx, outputs, x, ind, ind_inv, dim):
+        ctx.save_for_backward(ind, ind_inv)
+        ctx.dim = dim
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        ind, ind_inv = ctx.saved_tensors
+        result = NumpyTake.apply(grad_output, ind_inv, ind, ctx.dim)
+        return result, None, None, None
+
+
+class Select(torch.autograd.Function):
+    @staticmethod
+    def forward(x, idx):
+        return x[idx]
+
+    @staticmethod
+    def setup_context(ctx, outputs, x, idx):
+        ctx.x_shape = x.shape
+        ctx.idx = idx
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        result = grad_output.new_zeros(ctx.x_shape)
+        result[ctx.idx] = grad_output
+        return result, None
+
+
+def inplace_on_select_base(x):
+    y = x.clone()
+    y0 = Select.apply(x, 0)
+    y.sin_()
+    return y0
+
+
+def sample_inputs_inplace_on_select_base(opinfo, device, dtype, requires_grad, **kwargs):
+    make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+    yield SampleInput(make_arg(3, 5), args=())
+
+
+autograd_function_db = [
+    OpInfo(
+        'NumpyCubeAutogradFunction',
+        op=NumpyCube.apply,
+        supports_forward_ad=False,
+        supports_fwgrad_bwgrad=False,
+        sample_inputs_func=sample_inputs_numpy_cube,
+        dtypes=all_types_and(torch.bool, torch.half),
+        supports_out=False,
+    ),
+    OpInfo(
+        'NumpyExpAutogradFunction',
+        op=lambda x: NumpyExp_.apply(x.clone()),
+        inplace_variant=NumpyExp_.apply,
+        supports_forward_ad=False,
+        supports_fwgrad_bwgrad=False,
+        sample_inputs_func=sample_inputs_numpy_cube,
+        dtypes=all_types_and(torch.bool, torch.half),
+        supports_out=False,
+    ),
+    OpInfo(
+        'NumpyMulAutogradFunction',
+        op=NumpyMul.apply,
+        supports_forward_ad=False,
+        supports_fwgrad_bwgrad=False,
+        sample_inputs_func=sample_inputs_numpy_mul,
+        dtypes=all_types_and(torch.bool, torch.half),
+        supports_out=False,
+    ),
+    OpInfo(
+        'NumpyCubeNotComposableAutogradFunction',
```

Comment:
Shouldn't this one error when we run it with higher order gradients? Since we are testing with test_vjpvjp above, should that have a failure for this one?

Reply:
Computing with pytorch autograd:
- the first order gradient doesn't require grad
- (https://gist.github.com/zou3519/40ba6fcf1dc9e4140c20841d5f433d85)

The behavior of functorch transforms on trying to compute the gradient of a quantity that doesn't require grad is that it returns a Tensor full of zeros:
```
import torch
from functorch import grad
def f(x):
  return torch.zeros_like(x)

x = torch.tensor(1.)
grad(f)(x)  # torch.tensor(0)
```

the `test_vjpvjp` test is comparing vjp+vjp to vjp+{slightly modified PyTorch autograd}, so, both results end up giving us zero.

If we want to do something better here, we either need to change how functorch's grad transform works in general (or add some options to it), or improve the ctx.once_differentiable decorator. From my testing it looks like in regular PyTorch autograd, ctx.once_differentiable causes the output to not require_grad, but it would be better if it somehow attached a CantComputeGradients grad_fn to the output.

Reply:
From reading https://github.com/pytorch/pytorch/blob/master/torch/autograd/function.py#L422-L430

It looks like it already does attach the error grad_fn to backward, but it needs at least one of the inputs to backward to require grad. Taking the code example, if we pass in a grad_out that requires_grad that seems to produce the desired error.

```python
import torch

def to_numpy(x):
    return x.numpy()

class NumpyCubeNotComposable(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input):
        input_np = to_numpy(input)

        ctx.input_np = input_np
        ctx.device = input.device

        return torch.tensor(input_np ** 3, device=input.device), input_np

    @staticmethod
    @torch.autograd.function.once_differentiable
    def backward(ctx, grad_output, grad_saved):
        result_np = 3 * (ctx.input_np ** 2)
        return torch.tensor(result_np, device=ctx.device)

x = torch.randn([], requires_grad=True)
y, _ = NumpyCubeNotComposable.apply(x)
grad_y = torch.rand_like(y, requires_grad=True)
gx, = torch.autograd.grad(y, x, create_graph=True, grad_outputs=grad_y)

gx.backward() # RuntimeError: trying to differentiate twice a function that was marked with @once_differentiable
```

Reply:
Yeah, I can add a one-off test for this

Reply:
Fun fact: once_differentiable doesn't quite work with functorch grad. The problem is that it uses torch._C._functions.DelayedError, and DelayedError is a torch::autograd::Node that doesn't handle functorch.

I've brainstormed two potential solutions:
1. make a DelayedError autograd.Function, use that instead of  torch._C._functions.DelayedError when functorch is active
2. turn DelayedError into a ATen operation. The forward pass is an identity, while the backward pass raises an error.

Thoughts? (2) seems better for performance (it involves less overhead)

Reply:
Either sounds good to me! Maybe just the easier option for now, and leave (2) as a follow up if perf becomes an issue?

Reply:
Punting resolution of this to follow-up since there is more discussion to be had. For now I added a one-off test with an expectedFailure on it